### PR TITLE
feat: LT-1519 - HMRC Query

### PR DESCRIPTION
**Pull request by jfaracik**

### CODE REVIEW

File: applications/managers.py [hunk 1]
```diff
@@ -2,7 +2,7 @@
 
 
 class BaseApplicationManager(models.Manager):
-    def draft(self, organisation):
+    def drafts(self, organisation):
```
Comment (andytaylorkainos) [inline applications/managers.py]: should it not be `draft` as we are referring to an `application` of `draft` state? (one application)

Comment (jfaracik) [reply applications/managers.py]: `drafts` returns all drafts, not just a single application

File: cases/migrations/0003_notification_case_activity.py [hunk 2]
```diff
@@ -16,4 +16,4 @@ class Migration(migrations.Migration):
             name='case_activity',
             field=models.ForeignKey(null=True, on_delete=django.db.models.deletion.CASCADE, to='cases.CaseActivity'),
         ),
-    ]
+    ]
```
Comment (andytaylorkainos) [inline cases/migrations/0003_notification_case_activity.py]: new line

**Comment by jfaracik** [comment]
> Can you squash all the migration files made into one?

We will do!

File: conf/decorators.py [hunk 3]
```diff
@@ -35,9 +35,11 @@ def decorator(func):
         def inner(request, *args, **kwargs):
             application = _get_application(request, kwargs)
 
-            if application.licence_type != licence_type:
-                return JsonResponse(data={'errors': [f'This operation can only be used on applications of type '
-                                                     f'`{licence_type}`']},
+            types = [application_types] if isinstance(application_types, str) else application_types
```
Comment (andytaylorkainos) [inline conf/decorators.py]: either be explicit on the typing for the method or create one that handles the other type (application_types would suggest it should always receive  list, even if its a list of 1)

File: conf/decorators.py [hunk 4]
```diff
@@ -84,7 +86,10 @@ def inner(request, *args, **kwargs):
 
             if isinstance(request.request.user, ExporterUser):
                 application = _get_application(request, kwargs)
-                if application.organisation.id != request.request.user.organisation.id:
+                if application.application_type == ApplicationType.HMRC_QUERY and \
```
Comment (andytaylorkainos) [inline conf/decorators.py]: can you please but parenthesis around the two things being evaluated:
```(condition 1 and 2) or (condition 3 and 4)```

File: organisations/views/main.py [hunk 5]
```diff
@@ -43,24 +45,25 @@ def post(self, request):
...
-            serializer = OrganisationCreateSerializer(data=data)
-
-            if serializer.is_valid():
-                serializer.save()
-                return JsonResponse(data={'organisation': serializer.data},
-                                    status=status.HTTP_201_CREATED)
-
-            return JsonResponse(data={'errors': serializer.errors},
-                                status=status.HTTP_400_BAD_REQUEST)
+        data = JSONParser().parse(request)
```
Comment (andytaylorkainos) [inline organisations/views/main.py]: do we need to parse the request?

File: organisations/views/main.py [hunk 6]
```diff
@@ -43,24 +45,25 @@ def post(self, request):
...
+
+        if data.get('type') == 'individual':
+            try:
+                data['name'] = data['user']['first_name'] + " " + data['user']['last_name']
+            except AttributeError:
+                pass
+            except KeyError:
```
Comment (andytaylorkainos) [inline organisations/views/main.py]: can you combine the two errors
```
try:
   ...
except (error1, error2):
   pass
```

File: goodstype/models.py [hunk 7]
```diff
@@ -13,7 +14,7 @@ class GoodsType(models.Model):
     control_code = models.TextField(default=None, blank=True, null=True)
     is_good_end_product = models.BooleanField(default=None, blank=True, null=True)
     limit = models.Q(app_label='applications', model='application')
-    application = models.ForeignKey(OpenApplication, on_delete=models.CASCADE, related_name='open_application',
+    application = models.ForeignKey(BaseApplication, on_delete=models.CASCADE, related_name='base_application',
```
Comment (andytaylorkainos) [inline goodstype/models.py]: does this mean goodstypes are now applicable to standard applications?

File: applications/serializers/open_application.py [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+        ]
+
+    def update(self, instance, validated_data):
+        instance.have_you_been_informed = validated_data.get('have_you_been_informed', instance.have_you_been_informed)
+        if instance.have_you_been_informed == 'yes':
+            instance.reference_number_on_information_form = validated_data.get(
+                'reference_number_on_information_form', instance.reference_number_on_information_form)
+        else:
+            instance.reference_number_on_information_form = None
+        instance.last_modified_at = datetime.now(timezone.utc)
```
Comment (andytaylorkainos) [inline applications/serializers/open_application.py]: use django inbuilt timezone please

File: conf/helpers.py [hunk 9]
```diff
@@ -34,3 +34,7 @@ def replace_default_string_for_form_select(data, fields, replacement_value=None,
             if data[field] == default_select_value:
                 data[field] = replacement_value
     return data
+
+
+def get_value_from_enum(enum, value):
+    return [x for x in enum.choices if x[0] == value][0][1]
```
Comment (andytaylorkainos) [inline conf/helpers.py]: please dont iterate through all the choices if we dont need to

(psuedo)
```
for choice in enum.choices:
   if choice[0] == value:
      return choice[1]
```

File: goodstype/models.py (hunk 7 above)
Comment (jfaracik) [reply goodstype/models.py]: Yep!

File: organisations/views/main.py (hunk 6 above)
Comment (jfaracik) [reply organisations/views/main.py]: ABSOLUTELY

File: organisations/views/main.py (hunk 5 above)
Comment (jfaracik) [reply organisations/views/main.py]: Running it through json.loads now. We have to sadly as we modify the contents of it

File: applications/serializers/denial_reasons.py [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+    reasons = DenialReasonSerializer(read_only=False, many=True)
+
+    class Meta:
+        model = ApplicationDenialReason
+        fields = ('id',
+                  'reason_details',
+                  'reasons',)
+
+
+class ApplicationDenialReasonSerializer(serializers.ModelSerializer):
```
[the remark this answers is not in the document]
Comment (jfaracik) [reply applications/serializers/denial_reasons.py]: Good spot!

File: test_helpers/clients.py [hunk 11]
```diff
@@ -471,6 +492,35 @@ def create_open_application(self, organisation: Organisation, reference_name='Op
...
+    def create_hmrc_query(self, organisation: Organisation, reference_name='HMRC Query', safe_document=True):
+        application = HmrcQuery(name=reference_name,
+                                application_type=ApplicationType.HMRC_QUERY,
+                                activity='Trade',
+                                usage='Trade',
+                                organisation=organisation,
+                                hmrc_organisation=self.hmrc_organisation,
+                                end_user=self.create_end_user('End User', organisation),
+                                consignee=self.create_consignee('Consignee', organisation),
+                                reasoning='Because the exporter "forgot" to mention the nuclear submarine in his trunk')
```
[the remark this answers is not in the document]
Comment (jfaracik) [reply test_helpers/clients.py]: Aha, that will be fixed.

File: applications/migrations/0003_auto_20191030_1106.py [hunk 12]
```diff
@@ -0,0 +1,34 @@
+# Generated by Django 2.2.4 on 2019-10-30 11:06
+import django
```
[the remark this answers is not in the document]
Comment (jfaracik) [reply applications/migrations/0003_auto_20191030_1106.py]: We will be squashing at the end.

File: applications/serializers/generic_application.py [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+    def get_export_type(self, instance):
+        instance = get_application(instance.pk)
+        if hasattr(instance, 'export_type'):
+            return {
+                'key': instance.export_type,
+                'value': get_value_from_enum(ApplicationExportType, instance.export_type)
+            }
+        return None
+
+    def get_status(self, instance):
```
[the remark this answers is not in the document]
Comment (jfaracik) [reply applications/serializers/generic_application.py]: I don't think we can rename these as that'd affect the json it returns, eg renaming status to status_dict would return 'status_dict': { ... }

File: applications/urls.py [hunk 14]
```diff
@@ -10,152 +10,156 @@
     # ex: /applications/?submitted=true - List only submitted applications
     path(
         route='',
-        view=views.ApplicationList.as_view(),
+        view=applications.ApplicationList.as_view(),
         name='applications'
     ),
     # ex: /applications/<uuid:pk>/ - View an application
```
[the remark this answers is not in the document]
Comment (jfaracik) [reply applications/urls.py]: Good spot

File: applications/creators.py [hunk 15]
```diff
@@ -111,9 +134,13 @@ def validate_application_ready_for_submission(application):
...
-        validate_standard_licence(application, errors)
+    if application.application_type == ApplicationType.STANDARD_LICENCE:
+        _validate_standard_licence(application, errors)
+    elif application.application_type == ApplicationType.OPEN_LICENCE:
+        _validate_open_licence(application, errors)
+    elif application.application_type == ApplicationType.HMRC_QUERY:
+        _validate_hmrc_query(application, errors)
     else:
-        validate_open_licence(application, errors)
+        errors['unsupported_application'] = 'You can only validate a supported application type'
```
[the remark this answers is not in the document]
Comment (jfaracik) [reply applications/creators.py]: I don't think these messages will be there in the future, so we'll hold off for now.

File: applications/libraries/document_helpers.py [hunk 16]
```diff
@@ -103,3 +105,42 @@ def delete_party_document(party, application, user):
                                          party.name, user, application)
 
     return HttpResponse(status=status.HTTP_204_NO_CONTENT)
+
+
+def get_goods_type_document(goods_type):
+    if not goods_type:
+        return JsonResponse(data={'error': 'No such goods type'}, status=status.HTTP_400_BAD_REQUEST)
+
+    documents = GoodsTypeDocument.objects.filter(goods_type=goods_type)
```
[the remark this answers is not in the document]
Comment (jfaracik) [reply applications/libraries/document_helpers.py]: One document per goods type

[the remark this answers is not in the document]
Comment (jfaracik) [reply applications/libraries/document_helpers.py]: Good spot

File: applications/views/applications.py [hunk 17]
```diff
@@ -3,118 +3,128 @@
...
+            elif submitted:
+                applications = BaseApplication.objects. \
+                    submitted(organisation=self.request.user.organisation). \
+                    exclude(application_type=ApplicationType.HMRC_QUERY)
+            else:
+                applications = BaseApplication.objects. \
+                    drafts(organisation=self.request.user.organisation). \
+                    exclude(application_type=ApplicationType.HMRC_QUERY)
+
+        return applications
```
Comment (a-bek) [inline applications/views/applications.py]: `return applications.exclude(application_type=ApplicationType.HMRC_QUERY)`

File: applications/views/applications.py [hunk 18]
```diff
@@ -3,118 +3,128 @@
...
-        try:
-            submitted = optional_str_to_bool(request.GET.get('submitted'))
-        except ValueError as e:
-            return JsonResponse(data={'errors': str(e)}, status=status.HTTP_400_BAD_REQUEST)
-
-        if submitted is None:
-            qs = BaseApplication.objects.filter(organisation=request.user.organisation)
-        elif submitted:
-            qs = BaseApplication.objects.submitted(organisation=request.user.organisation)
+        if self.request.user.organisation.type == OrganisationType.HMRC:
```
Comment (a-bek) [inline applications/views/applications.py]: You could return applications here and remove `else`.

File: applications/helpers.py [hunk 19]
```diff
@@ -0,0 +1,45 @@
...
+from applications.serializers.hmrc_query import HmrcQueryCreateSerializer, HmrcQueryViewSerializer, \
+    HmrcQueryUpdateSerializer
+from applications.serializers.open_application import OpenApplicationCreateSerializer, OpenApplicationUpdateSerializer, \
+    OpenApplicationViewSerializer
+from applications.serializers.standard_application import StandardApplicationCreateSerializer, \
+    StandardApplicationUpdateSerializer, StandardApplicationViewSerializer
+from conf.exceptions import BadRequestError
+
+
+def get_application_view_serializer(application: BaseApplication):
```
Comment (a-bek) [inline applications/helpers.py]: This should just take `application_type` as argument like below?

File: goodstype/helpers.py [hunk 20]
```diff
@@ -8,3 +9,12 @@ def get_goods_type(pk):
         return GoodsType.objects.get(pk=pk)
     except GoodsType.DoesNotExist:
         raise Http404
+
+
+def delete_goods_type_document_if_exists(goods_type: GoodsType):
```
[the remark this answers is not in the document]
Comment (jfaracik) [reply goodstype/helpers.py]: delete_goods_type_document_if_exists handles itself failing better, but it isn't particularly pythonic. I'm happy to keep it for now in the interest of time but I too would like to get rid of it in the future.

File: applications/helpers.py (hunk 19 above)
Comment (jfaracik) [reply applications/helpers.py]: Only reason application_create takes application type is that the actual application isn't created at that time, and that's the only piece of information we have about it.

File: applications/serializers/generic_application.py [hunk 21]
```diff
@@ -0,0 +1,150 @@
...
+    def get_case(self, instance):
+        try:
+            return Case.objects.get(application=instance).id
+        except Case.DoesNotExist:
+            # Case will only exist if application has been submitted
+            return None
+
+    class Meta:
+        model = BaseApplication
+        fields = [
```
Comment (andytaylorkainos) [inline applications/serializers/generic_application.py]: please use tuples. we use them everywhere else for fields

File: applications/serializers/open_application.py [hunk 22]
```diff
@@ -0,0 +1,105 @@
...
+
+
+class OpenApplicationCreateSerializer(GenericApplicationCreateSerializer):
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        self.initial_data['organisation'] = self.context.id
+
+    class Meta:
+        model = OpenApplication
+        fields = [
```
[the remark this answers is not in the document]
Comment (andytaylorkainos) [reply applications/serializers/open_application.py]: the only reason at the mo is consistency

File: applications/serializers/generic_application.py (hunk 21 above)
Comment (jfaracik) [reply applications/serializers/generic_application.py]: You can't concatenate tuples, hence the lists.

File: applications/tests/tests_end_user.py [hunk 23]
```diff
@@ -1,7 +1,8 @@
+from unittest import mock
```
[the remark this answers is not in the document]
Comment (jfaracik) [reply applications/tests/tests_end_user.py]: Yes, for deleting documents.

File: applications/views/documents.py [hunk 24]
```diff
@@ -64,3 +69,41 @@ def delete(self, request, application, doc_pk):
...
+        request_body=GoodsTypeDocumentSerializer,
+        responses={
+            400: 'JSON parse error'
+        })
+    @transaction.atomic
+    @allowed_application_types(ApplicationType.HMRC_QUERY)
+    @authorised_users(ExporterUser)
+    def delete(self, request, application, goods_type_pk):
+        goods_type = get_goods_type(goods_type_pk)
+        return delete_goods_type_document(goods_type)
```
[the remark this answers is not in the document]
Comment (jfaracik) [reply applications/views/documents.py]: @markm5A8Z9F43913A

**Pull request merged**